File: week-2/problem-set/prob11.py
"""Prob11 example.

This file belongs to the week 2 data-structure and graph practice solutions. It uses stock_span to illustrate the main idea with a small runnable example.
"""

def stock_span(prices):
    """Compute the stock span for each day using a stack.

    Args:
        prices (list): The stock prices used by the span calculation.

    Returns:
        list: The list produced by the algorithm.
    """
    stk = []
    span = []

    for i in range(len(prices)):
        cur_span = 1
        while stk and prices[stk[-1]] <= prices[i]:
            prev_idx = stk.pop()
            cur_span += span[prev_idx]
        span.append(cur_span)
        stk.append(i)

    return span



import time

start_time = time.time()
# calculateShares(10**12)
# calculateShares(3)
prices = [100, 80, 60, 70, 60, 75, 85]
print(stock_span(prices))
end_time = time.time()
print(end_time - start_time)

n = 10**12
print(n)

File: week-2/problem-set/prob11.cpp
/**
 * @file prob11.cpp
 * @brief C++ version of the stock-span example.
 */

#include <chrono>
#include <iostream>
#include <vector>

using namespace std;
using Clock = chrono::high_resolution_clock;

/**
 * @brief Compute the stock span for each day.
 * @param prices const vector<int>& Daily stock prices.
 * @return vector<int> Span values where each entry counts consecutive earlier days with price less than or equal to the current day.
 */
vector<int> stock_span(const vector<int>& prices) {
    vector<int> stack_indices;
    vector<int> span;

    for (int i = 0; i < static_cast<int>(prices.size()); ++i) {
        int cur_span = 1;
        while (!stack_indices.empty() && prices[stack_indices.back()] <= prices[i]) {
            int prev_index = stack_indices.back();
            stack_indices.pop_back();
            cur_span += span[prev_index];
        }
        span.push_back(cur_span);
        stack_indices.push_back(i);
    }

    return span;
}

/**
 * @brief Print a vector of integers in bracket form.
 * @param values const vector<int>& Vector to print. Each element has type int.
 * @return void This helper only prints data and does not return a value.
 */
void print_vector(const vector<int>& values) {
    cout << '[';
    for (size_t index = 0; index < values.size(); ++index) {
        cout << values[index];
        if (index + 1 != values.size()) {
            cout << ", ";
        }
    }
    cout << ']';
}

/**
 * @brief Run the stock-span demonstration and print the runtime.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    vector<int> prices = {100, 80, 60, 70, 60, 75, 85};

    auto start_time = Clock::now();
    vector<int> span = stock_span(prices);
    auto end_time = Clock::now();

    print_vector(span);
    cout << '\n';
    chrono::duration<double> elapsed = end_time - start_time;
    cout << elapsed.count() << '\n';

    long long n = 1000000000000LL;
    cout << n << '\n';
    return 0;
}
